x/auth/keeper/migrations.go

```diff
@@ -0,0 +1,146 @@
+package keeper
+
+import (
+	"fmt"
+
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+
+	"github.com/cosmos/cosmos-sdk/x/auth/vesting/exported"
+
+	"github.com/gogo/protobuf/grpc"
+	"github.com/gogo/protobuf/proto"
+	abci "github.com/tendermint/tendermint/abci/types"
+
+	"github.com/cosmos/cosmos-sdk/baseapp"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/auth/types"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+)
+
+const (
+	delegatorDelegationPath = "/cosmos.staking.v1beta1.Query/DelegatorDelegations"
+	balancesPath            = "/cosmos.bank.v1beta1.Query/AllBalances"
+)
+
+// Migrator is a struct for handling in-place store migrations.
+type Migrator struct {
+	keeper      AccountKeeper
+	queryServer grpc.Server
+}
+
+// NewMigrator returns a new Migrator.
+func NewMigrator(keeper AccountKeeper, queryServer grpc.Server) Migrator {
+	return Migrator{keeper: keeper, queryServer: queryServer}
+}
+
+// Migrate1to2 migrates from version 1 to 2.
+func (m Migrator) Migrate1to2(ctx sdk.Context) error {
+	m.keeper.IterateAccounts(ctx, func(account types.AccountI) (stop bool) {
+		asVesting := vesting(account)
+		if asVesting == nil {
+			return false
+		}
+
+		addr := account.GetAddress().String()
+		balance := getBalance(
+			ctx,
+			addr,
+			m.queryServer,
+		)
+
+		delegations := getDelegatorDelegations(
+			ctx,
+			addr,
+			m.queryServer,
+		)
+
+		if delegations.IsAllGTE(asVesting.GetOriginalVesting()) {
+			delegations = asVesting.GetOriginalVesting()
+		}
+
+		if balance.IsAllLT(delegations) {
+			balance = balance.Add(delegations...)
+		}
+
+		asVesting.TrackDelegation(ctx.BlockTime(), balance, delegations)
+
+		m.keeper.SetAccount(ctx, account)
+
+		return false
+	})
+
+	return nil
+}
+
+func vesting(account types.AccountI) exported.VestingAccount {
+	v, ok := account.(exported.VestingAccount)
+	if !ok {
+		return nil
+	}
+
+	return v
+}
+
+func getDelegatorDelegationsSum(ctx sdk.Context, address string, queryServer grpc.Server) sdk.Coins {
+	querier, ok := queryServer.(*baseapp.GRPCQueryRouter)
+	if !ok {
+		panic(fmt.Sprintf("unexpected type: %T wanted *baseapp.GRPCQueryRouter", queryServer))
+	}
+
+	queryFn := querier.Route(delegatorDelegationPath)
+
+	q := &stakingtypes.QueryDelegatorDelegationsRequest{
+		DelegatorAddr: address,
+	}
+
+	b, err := proto.Marshal(q)
+	if err != nil {
+		panic(err)
+	}
+	req := abci.RequestQuery{
+		Data: b,
+		Path: delegatorDelegationPath,
+	}
+	resp, err := queryFn(ctx, req)
```

Comment:
We're going through ABCI just to query other keepers' state. Surely there's a better way to achieve that.

Some things that are ugly because of this strategy:
- we need to hardcode the gRPC paths on the top of this file
- `proto.Unmarshal()` is dangerously called multiple times, without the interfaceRegistry

The way we do inter-module keeper querying right now is via exported keeper interfaces. Maybe we should keep like that? I see the challenge here is that the AuthKeeper does not have access to BankKeeper & StakingKeeper.


I have a proposal, but not sure if it suits you: Put the migration logic inside the x/staking keeper. That's because x/staking has a reference to both AuthKeeper and BankKeeper, and we should be able to handle migrations without manipulating any gRPC stuff, just by calling functions.

(ADR-033 would elegantly solve this, btw)

Reply:
> Put the migration logic inside the x/staking keeper. That's because x/staking has a reference to both AuthKeeper and BankKeeper, and we should be able to handle migrations without manipulating any gRPC stuff, just by calling functions.

I agree on the gRPC part, this is not optimal for sure.

I don't feel like having this migration in `staking` because we're touching `auth`'s state - while doing so removes gRPC inconveniences it's not very idiomatic nor correct IMO.

> (ADR-033 would elegantly solve this, btw)

While writing this we often invoked some sort of way for doing inter-module communication :-)